File: app/src/main/java/br/com/algorit/mangaalert/roomdatabase/repository/NovelRepository.java
package br.com.algorit.mangaalert.roomdatabase.repository;

import android.app.Application;
import android.os.AsyncTask;

import androidx.lifecycle.LiveData;

import java.util.List;

import br.com.algorit.mangaalert.roomdatabase.dao.NovelDao;
import br.com.algorit.mangaalert.roomdatabase.database.AlertRoomDatabase;
import br.com.algorit.mangaalert.roomdatabase.model.Novel;

public class NovelRepository {
    private final NovelDao novelDao;
    private final LiveData<List<Novel>> allNovels;

    public NovelRepository(Application application) {
        AlertRoomDatabase database = AlertRoomDatabase.getDatabase(application);
        novelDao = database.novelDao();
        allNovels = novelDao.getAll();
    }

    public LiveData<List<Novel>> getAllNovels() {
        return allNovels;
    }

    public void insert(Novel novel) {
        new InsertAsyncTask(novelDao).execute(novel);
    }

    public void updateCapitulo(Novel novel) {
        new UpdateAsyncTask(novelDao, false).execute(novel);
    }

    public void updateChecked(Novel novel) {
        new UpdateAsyncTask(novelDao, true).execute(novel);
    }

    public void uncheckAll() {
        new UncheckAllAsyncTask(novelDao).execute();
    }

    private static class InsertAsyncTask extends AsyncTask<Novel, Void, Void> {
        private final NovelDao asyncTaskDao;

        InsertAsyncTask(NovelDao dao) {
            asyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(final Novel... novels) {
            Novel novel = novels[0];
            if (asyncTaskDao.findByNome(novel.getNome()) != null)
                new UpdateAsyncTask(asyncTaskDao, false).execute(novels);
            else asyncTaskDao.insert(novel);
            return null;
        }
    }

    private static class UpdateAsyncTask extends AsyncTask<Novel, Void, Void> {
        private final NovelDao asyncTaskDao;
        private final boolean updateChecked;

        UpdateAsyncTask(NovelDao dao, boolean updateChecked) {
            asyncTaskDao = dao;
            this.updateChecked = updateChecked;
        }

        @Override
        protected Void doInBackground(final Novel... novels) {
            Novel novel = novels[0];
            if (updateChecked) {
                asyncTaskDao.updateChecked(novel.isChecked(), novel.getNome());
            } else {
                asyncTaskDao.updateCapitulo(novel.getCapitulo(), novel.getNome());
            }
            return null;
        }
    }

    private static class UncheckAllAsyncTask extends AsyncTask<Void, Void, Void> {
        private final NovelDao asyncTaskDao;

        UncheckAllAsyncTask(NovelDao dao) {
            this.asyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(Void... voids) {
            asyncTaskDao.uncheckAll();
            return null;
        }
    }
}

File: app/src/main/java/br/com/algorit/mangaalert/roomdatabase/repository/NovelRepository.kt
package br.com.algorit.mangaalert.roomdatabase.repository

import android.app.Application
import android.os.AsyncTask
import androidx.lifecycle.LiveData
import br.com.algorit.mangaalert.roomdatabase.dao.NovelDao
import br.com.algorit.mangaalert.roomdatabase.database.AlertRoomDatabase.Companion.getDatabase
import br.com.algorit.mangaalert.roomdatabase.model.Novel

class NovelRepository(application: Application?) {
    private val novelDao: NovelDao
    val allNovels: LiveData<List<Novel>>

    init {
        val database = getDatabase(application!!)
        novelDao = database!!.novelDao()
        allNovels = novelDao.all
    }

    fun insert(novel: Novel) {
        if (novelDao.findByNome(novel.nome) != null)
            UpdateAsyncTask(novelDao, false).execute(novel)
        else
            InsertAsyncTask(novelDao).execute(novel)
    }

    fun updateCapitulo(novel: Novel) {
        UpdateAsyncTask(novelDao, false).execute(novel)
    }

    fun updateChecked(novel: Novel) {
        UpdateAsyncTask(novelDao, true).execute(novel)
    }

    fun uncheckAll() {
        UncheckAllAsyncTask(novelDao).execute()
    }

    private class InsertAsyncTask(private val asyncTaskDao: NovelDao) :
        AsyncTask<Novel, Void, Void>() {
        override fun doInBackground(vararg novels: Novel): Void? {
            val novel = novels[0]
            asyncTaskDao.insert(novel)
            return null
        }
    }

    private class UpdateAsyncTask(
        private val asyncTaskDao: NovelDao,
        private val updateChecked: Boolean
    ) : AsyncTask<Novel, Void, Void>() {
        override fun doInBackground(vararg novels: Novel): Void? {
            val novel = novels[0]
            if (updateChecked) {
                asyncTaskDao.updateChecked(novel.isChecked, novel.nome)
            } else {
                asyncTaskDao.updateCapitulo(novel.capitulo!!, novel.nome)
            }
            return null
        }
    }

    private class UncheckAllAsyncTask(private val asyncTaskDao: NovelDao) :
        AsyncTask<Void, Void, Void>() {
        override fun doInBackground(vararg voids: Void): Void? {
            asyncTaskDao.uncheckAll()
            return null
        }
    }
}